UI support for openroberta

ensonic opened the pull request.

ensonic committed:
data: copy over the setup for generating png files from pbm

This is a preparation to add more images in future changes.

ensonic committed:
Add ui support for lab.openroberta.org

The ui contains of a dialog to connect to the server and status icon. The actual
communication with the server is done by a system service that exposes a bus
control interface.

dlech commented on src/controller/OpenRobertaController.vala:
```diff
@@ -0,0 +1,184 @@
+        }
+
+        void on_status_changed (string message) {
+            string[] offline = {"connected", "disconnected" };
```
nit: space between `{` and `"`

ensonic replied:
done

dlech commented on src/controller/OpenRobertaController.vala:
```diff
@@ -0,0 +1,184 @@
+                open_roberta_window.connected = true;
+                if (message == "registered") {
+                    if (pin_dialog != null) {
+                        info ("connection established, closing the dialog");
```
This sounds like it should be `debug` instead of `info`.

ensonic replied:
done

dlech commented:
I'm trying this out and I have a few questions (this is all in the desktop test version, so the real controller may be different):
1. When the OpenRoberta service is not available, there is nothing on the screen:
   
   ![empty openroberta window screenshot](https://cloud.githubusercontent.com/assets/963645/10146213/3447be90-65ed-11e5-9e4e-71f7943f5b3d.png)
   
   I know that you would like to remove the menu item so that you can't get here if the service is not running, but it might be nice to have a "service not running" message here just in case.
2. Should the buttons here be radio buttons?
   
   ![desktop test buttons screenshot](https://cloud.githubusercontent.com/assets/963645/10146324/e6a9e8b0-65ed-11e5-868b-50bd1e79dccc.png)
3. Regarding the above, I don't see a difference between connect and disconnect or registered and executing
4. I take it that this screen is asking the user to select which server to connect to? If so, it could use a message saying so to make it obvious.
   
   ![connect screen](https://cloud.githubusercontent.com/assets/963645/10146401/4405d410-65ee-11e5-98ff-c200a9d89cb7.png)
   
   There also seems to be a menu item with no text between "10.0.1.10:1990" and "Another...". Looking at the source code, this is a "user-defined" item. It should not be in the menu until it is defined. Is there a place where these user-defined values will be stored?
5. We need to add the OpenRoberta service package to the dependencies in `debian/control`.

ensonic commented:
1) will do.
2/3) Those just fire the status changes, the dbus service would fire too. For the UI we only when we're registered with the cloud-ide or in the icon.
4) I can add a mesage - good idea. I just did a 1:1 from the lejos menu to get things up and running.
Also the user-define item is stored in /etc/openrobertalab.conf (keyfile). If the user-define item is not set, selecting it would also open the editor, but I agree it is not optimal - was just easier to code.
5.) The idea is that there is no dependency. If the dbus iface is not there, the menu item won't show up.

dlech commented:
2/3) The desktop test version does not need to actually emulate exactly how it will behave on the EV3. The point is to be able to create every possible value for anything that can be changed so that you can see how it looks (and also create faked screenshots for documentation).
5) Perhaps Recommends or Suggests would be appropriate then.

ensonic committed:
robertalab: use different info label when we are connected

ensonic committed:
robertalab: block the back-button-long-press handler

When executing our code, we are checking this button too and when we get back to
brickman, we don't expect to see the shutdown dialog. This happens since brickman
uses the hardware key to detect the long press.

ensonic committed:
openroberta: store selected server

When we re-enter the menu, we can preselect the previously used entry. We can
also show to what server we are connected to.

ensonic committed:
Only switch back to vt 1 when execution finished.

The openroberta service also sends updates and just unconditionally switching to
vt 1 would interfer with apps launched from brickman.

ensonic committed:
Update to new IBrickManagerModule api.

ensonic committed:
openrobertalab: remove local server

This was lejos specific. In our setup the brick won't have a specific ip when
connected via usb.

ensonic committed:
debian: add Suggests field to control

ensonic committed:
openrobertalab: simplify the fake controller

Only expose the states that we need to simulate the UI.

ensonic committed:
openrobertalab: add test code to edit the custom server

ensonic commented:
2/3) done: simplified the ui in the test code
5) done

dlech commented on data/pnmtopng.cmake:
```diff
@@ -0,0 +1,13 @@
+    OUTPUT_VARIABLE
+        output
+    ERROR_VARIABLE
+        error)
```
nit: newline at end of file

ensonic replied:
This is a copy from
https://github.com/ev3dev/ev3devKit/blob/ev3dev-jessie/data/pnmtopng.cmake
which neither has a newline at the end of the file. I've sent you a PR for that to keep it consistent.

Done.

dlech commented on src/controller/OpenRobertaController.vala:
```diff
@@ -0,0 +1,238 @@
+        MessageDialog pin_dialog;
+        bool executing_user_code = false;
+
+        public bool available { get; set; default = false; }
```
nit: don't need initializer. default value is already false.

dlech commented on src/controller/OpenRobertaController.vala:
```diff
@@ -0,0 +1,238 @@
+
+            notify["available"].connect (() => {
+                if (!available) {
+                    open_roberta_window.info.text =
```
It seems like if we are setting this here, we would need to set it to something else in the "else" block.

dlech replied:
I see where this is handled now. I'm thinking I need to add a new control to ev3devKit to handle this. Basically a Notebook without the tabs.

Bottom line, this is OK for now.

dlech commented on src/controller/OpenRobertaController.vala:
```diff
@@ -0,0 +1,238 @@
+                    open_roberta_window.info.text =
+                        "Service openrobertalab is not running.";
+                } else {
+                  open_roberta_window.notify_property ("connected");
```
nit: indent 4 spaces

dlech commented on src/controller/OpenRobertaController.vala:
```diff
@@ -0,0 +1,238 @@
+        }
+
+        void switch_to_program_screen () {
+          Posix.system ("/bin/chvt 2");
```
nit: indent 4 spaces

dlech commented on src/controller/OpenRobertaController.vala:
```diff
@@ -0,0 +1,238 @@
+        }
+
+        void switch_to_brickman_screen () {
+          Posix.system ("/bin/chvt 1");
```
nit: indent 4 spaces

dlech commented on src/controller/OpenRobertaController.vala:
```diff
@@ -0,0 +1,238 @@
+          GLib.SignalHandler.unblock (global_manager, handler_id);
+        }
+    }
+}
```
nit: newline at end of file

dlech commented on src/view/OpenRobertaWindow.vala:
```diff
@@ -0,0 +1,99 @@
+
+namespace BrickManager {
+    public class OpenRobertaWindow : BrickManagerWindow {
+        internal Label info;
```
nit: variable names should say what it is, so `info_label`, `public_server_menu_item`, etc.

ensonic replied:
called it status_info - info_label would repeat the type

dlech replied:
I get not including the type name in variable name just for the sake of including the type, but in the case of UI widgets I always include it because it describes what the variable is. For example, the `*_server` variables below do not represent a server, they represent a menu item, so I like the variable name to be `*_menu_item`.

dlech commented on src/view/OpenRobertaWindow.vala:
```diff
@@ -0,0 +1,99 @@
+        internal Ui.MenuItem config_custom;
+        internal Ui.MenuItem disconnect_server;
+
+        public bool connected { get; set; default = false; }
```
nit: don't need default initializers here.

dlech commented on src/view/OpenRobertaWindow.vala:
```diff
@@ -0,0 +1,99 @@
+
+        void on_connected_changed () {
+            menu.remove_all_menu_items ();
+            if (!connected) {
```
nit: when if statements have and else, leave out the `!` and swap the blocks.

dlech commented on src/controller/OpenRobertaController.vala:
```diff
@@ -0,0 +1,238 @@
+
+        void on_status_changed (string message) {
+            debug ("service status: '%s'", message);
+            string[] offline = { "connected", "disconnected" };
```
`connected` == offline ???

ensonic replied:
Added comments for the states. Changed it to online containing (registered, executing) and flipped the cond.

dlech commented on src/controller/OpenRobertaController.vala:
```diff
@@ -0,0 +1,238 @@
+                        font = BrickManagerWindow.big_font
+                    };
+                    pin_dialog = new MessageDialog.with_content ("Pairing code", label);
+                    pin_dialog.closed.connect (() => { pin_dialog = null; });
```
`pin_dialog` should be a local variable, then you don't need the `connect` here. As it is, you have a reference cycle, which will cause a memory leak.

ensonic replied:
Still need to to this ...

dlech replied:
I was thinking about this as I was falling asleep last night. If you make the variable declaration `weak MessageDialog pin_dialog`, this should fix the memory leak.

dlech replied:
then you don't have to use a local variable.

ensonic committed:
openrobertalab: code-review comment cleanups
